### PR TITLE
Docs: Add GraphQL docs

### PR DESCRIPTION
**Pull request by Rachael-Graham**
BOT NOTES: 
resolves https://github.com/solo-io/gloo-mesh-enterprise/issues/1242

### CODE REVIEW

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 1]
```diff
@@ -1,45 +1,75 @@
...
+API gateways solve the problem of exposing multiple microservices with differing implementations from a single location and scheme, and by talking to a single owner. GraphQL integrates well with API gateways by exposing your API without versioning and allowing clients to interact with backend APIs on their own terms. Additionally, you can mix and match your GraphQL graph with your existing REST routes to test GraphQL integration features and migrate to GraphQL at a pace that makes sense for your organization.
+
+Gloo Edge solves the problems that other API gateways face when exposing GraphQL services by allowing you to configure GraphQL at the route level. API gateways are often used to rate limit, authorize and authenticate, and inject other centralized edge networking logic at the route level. However, because most GraphQL servers are exposed as a single endpoint within an internal network behind API gateways, you cannot add route-level customizations. With Gloo Edge, route-level customization logic is embedded into the API gateway.
+
+## Step 1: Install GraphQL
+
+GraphQL resolution is an experimental feature included in Gloo Edge Enterprise version 1.10.0-beta1 and later.
```
Comment (Rachael-Graham) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:29]: Is this still the official release that graphQL support was included in, or should we change this to a later version?

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 2]
```diff
@@ -1,45 +1,75 @@
...
-within an internal network behind API gateways, you cannot add route-level customizations.
-With Gloo Edge, route-level customization logic is embedded into the API gateway.
-
-## Example: GraphQL with Gloo Edge
-
-**Before you begin**: Deploy the sample pet store application, which you will expose behind a GraphQL server embedded in Envoy.
-```shell
-kubectl apply -f https://raw.githubusercontent.com/solo-io/gloo/v1.2.9/example/petstore/petstore.yaml
+{{% notice warning %}}
+This feature is experimental. Do not use this feature in a production environment.
```
Comment (Rachael-Graham) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:14]: Is this line still correct? Should this read "alpha" or "beta" instead of experimental?

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 3]
```diff
@@ -52,57 +82,148 @@ spec:
...
    ```
 
-Remember that previously the REST request returned the following JSON to the GraphQL server:
-```json
-[{"id":1,"name":"Dog","status":"available"},{"id":2,"name":"Cat","status":"pending"}]
-```
-Data filtering is one advantage of using GraphQL instead of querying the upstream directly. Because the GraphQL query is issued for only the name of the pets, GraphQL is able to filter out any data in the response that is irrelevant to the query, and return only the data that is specifically requested.
+## Try it yourself
+
+TODO: need guidance on how to develop the `GraphQLSchema` CR for their own use cases and apps
```
Comment (Rachael-Graham) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:225]: should we provide some extra guidance for how to do this, or let the example (starting at line 142) suffice for now?

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 1 above)
Comment (kcbabo) [reply docs/content/guides/traffic_management/request_processing/graphql/_index.md:29]: Let's change that to 1.11.0-beta3 and later. Might as well go with the most recent version when we update the doc. We don't want people using this at all with 1.10.x.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 2 above)
Comment (kcbabo) [reply docs/content/guides/traffic_management/request_processing/graphql/_index.md:14]: Let's call it alpha.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 3 above)
Comment (kcbabo) [reply docs/content/guides/traffic_management/request_processing/graphql/_index.md:225]: The easiest thing might be to point at the graphql-bookinfo repo as a more in-depth example to model custom use cases after:
https://github.com/kcbabo/graphql-bookinfo

What we need to add before SoloCon is a section that explains the primary configuration points and workflow for deploying your own GraphQL API (i.e. one that doesn't leverage autodiscovery and registration).

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 4]
```diff
@@ -1,45 +1,75 @@
...
+This feature is experimental. Do not use this feature in a production environment.
+{{% /notice %}}
+
+## About
+
+### Why GraphQL?
+GraphQL is a server-side query language and runtime you can use to expose your APIs as an alternative to REST APIs. GraphQL allows you to request only the data you want and handle any subsequent requests on the server side, saving numerous expensive origin-to-client requests by instead handling requests in your internal network.
+
+### Using GraphQL in an API gateway
+API gateways solve the problem of exposing multiple microservices with differing implementations from a single location and scheme, and by talking to a single owner. GraphQL integrates well with API gateways by exposing your API without versioning and allowing clients to interact with backend APIs on their own terms. Additionally, you can mix and match your GraphQL graph with your existing REST routes to test GraphQL integration features and migrate to GraphQL at a pace that makes sense for your organization.
```
Comment (artberger) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:23]: I was having a hard time understanding this, so took a shot at shortening some of the sentences and adding more connector words.
```suggestion
API gateways expose microservices with different implementations from a single location and schema. The API gateway acts like a single owner for all requests and responses. As such, it can shape traffic according to consistent policies that you set. When you integrate with GraphQL, you get the benefits of an API gateway and more. GraphQL exposes your API without versioning and lets clients interact with the API on their own terms. Additionally, you can mix and match your GraphQL graph with your existing REST routes. This setup lets you test and migrate to GraphQL at a pace that makes sense for your organization.
```

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 5]
```diff
@@ -1,45 +1,75 @@
...
+## Step 1: Install GraphQL
+
+GraphQL resolution is an experimental feature included in Gloo Edge Enterprise version 1.10.0-beta1 and later.
+
+To try out GraphQL, install Gloo Edge in a development environment. Note that you currenty cannot update an existing installation to use GraphQL. Be sure to specify version 1.10.0-beta1 or later. For the latest available version, see the [Gloo Edge Enterprise changelog]({{% versioned_link_path fromRoot="/reference/changelog/enterprise/" %}}).
 ```
-Note that any `/GET` requests to `/api/pets` of this service return the following JSON output:
-```json
-[{"id":1,"name":"Dog","status":"available"},{"id":2,"name":"Cat","status":"pending"}]
+glooctl install gateway enterprise --version 1.11.0-beta3 --license-key=<LICENSE_LEY>
```
Comment (kcbabo) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:33]: We should mention that you need an Edge Enterprise license that specifically enables the GraphQL capability. Customers or trial evaluators should contact their account team for the appropriate license key.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 6]
```diff
@@ -1,45 +1,75 @@
...
 ```
 
-1. Create a virtual service that defines a `Route` with a `graphqlSchemaRef` as the
-destination. In this example, all traffic to `/graphql` is handled by the GraphQL server in the Envoy proxy. 
+Next, you can try out GraphQL filtering with sample apps such as [Pet Store](#pet-store) and [Bookinfo](#bookinfo).
```
Comment (kcbabo) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:36]: Not sure about the reference to 'filtering' here. Do you mean it in the sense of an Envoy filter? You could probably just delete that word altogether and the sentence would still read fine.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 7]
```diff
@@ -1,45 +1,75 @@
...
+   kubectl apply -f https://raw.githubusercontent.com/solo-io/gloo/master/example/petstore/petstore.yaml
+   ```
+   Note that any `/GET` requests to `/api/pets` of this service return the following JSON output:
+   ```json
+   [{"id":1,"name":"Dog","status":"available"},{"id":2,"name":"Cat","status":"pending"}]
+   ```
+
+2. To allow Gloo Edge to automatically discover API specifications and create GraphQL schemas, turn on FDS discovery.
+   ```sh
+   kubectl patch settings -n gloo-system default --type=merge --patch '{"spec":{"discovery":{"fdsMode":"BLACKLIST"}}}'
```
Comment (kcbabo) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:53]: We should talk to Sai about whether we should lock this down a bit more. Maybe whitelist a specific deployment or namespace instead of everything?
@saiskee wdyt?

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 8]
```diff
@@ -1,45 +1,75 @@
...
+
+### Using GraphQL in an API gateway
+API gateways solve the problem of exposing multiple microservices with differing implementations from a single location and scheme, and by talking to a single owner. GraphQL integrates well with API gateways by exposing your API without versioning and allowing clients to interact with backend APIs on their own terms. Additionally, you can mix and match your GraphQL graph with your existing REST routes to test GraphQL integration features and migrate to GraphQL at a pace that makes sense for your organization.
+
+Gloo Edge solves the problems that other API gateways face when exposing GraphQL services by allowing you to configure GraphQL at the route level. API gateways are often used to rate limit, authorize and authenticate, and inject other centralized edge networking logic at the route level. However, because most GraphQL servers are exposed as a single endpoint within an internal network behind API gateways, you cannot add route-level customizations. With Gloo Edge, route-level customization logic is embedded into the API gateway.
```
Comment (artberger) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:25]:
```suggestion
Gloo Edge extends API gateway and GraphQL capabilities with route-level control.  Usually, API gateways apply edge networking logic at the route level. For example, the gateway might rate limit, authorize, and authenticate requests. Most GraphQL servers are a separate endpoint behind the API gateway. Therefore, you cannot add route-level customizations. In contrast, Gloo Edge embeds route-level customization logic into the API gateway.
```

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 9]
```diff
@@ -52,57 +82,148 @@ spec:
...
+```
+Data filtering is one advantage of using GraphQL instead of querying the upstream directly. Because the GraphQL query is issued for only the name of the pets, GraphQL is able to filter out any data in the response that is irrelevant to the query, and return only the data that is specifically requested.
+
+## Step 3: GraphQL resolvers with Bookinfo {#bookinfo}
+
+Next, explore GraphQL resolution with the Bookinfo sample application.
+
+In Gloo Edge, you can create GraphQL resolvers to fetch the data from your backend. Today Gloo Edge supports REST and gRPC resolvers. In the following steps, you create resolvers that point to Bookinfo services and use the resolvers in a GraphQL schema.
+
+1. Start by installing the Istio sample app, Bookinfo.
```
Comment (kcbabo) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:114]: You don't need to install Istio -- we're just using the Bookinfo example app from the Istio distro.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 1 above)
Comment (artberger) [reply docs/content/guides/traffic_management/request_processing/graphql/_index.md:29]: Since 1.10 is officially released, I would recommend users starting with that vs. a beta version.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 10]
```diff
@@ -1,45 +1,75 @@
...
+
+Gloo Edge solves the problems that other API gateways face when exposing GraphQL services by allowing you to configure GraphQL at the route level. API gateways are often used to rate limit, authorize and authenticate, and inject other centralized edge networking logic at the route level. However, because most GraphQL servers are exposed as a single endpoint within an internal network behind API gateways, you cannot add route-level customizations. With Gloo Edge, route-level customization logic is embedded into the API gateway.
+
+## Step 1: Install GraphQL
+
+GraphQL resolution is an experimental feature included in Gloo Edge Enterprise version 1.10.0-beta1 and later.
+
+To try out GraphQL, install Gloo Edge in a development environment. Note that you currenty cannot update an existing installation to use GraphQL. Be sure to specify version 1.10.0-beta1 or later. For the latest available version, see the [Gloo Edge Enterprise changelog]({{% versioned_link_path fromRoot="/reference/changelog/enterprise/" %}}).
```
Comment (artberger) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:31]: Is this still true? I'm wondering if it should be a note in the 1.10 Upgrade doc, if so: https://docs.solo.io/gloo-edge/latest/operations/upgrading/1.10.0/

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 5 above)
Comment (artberger) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:33]: This says 1.11, probably change to 1.10.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 11]
```diff
@@ -52,57 +82,148 @@ spec:
...
+   ```sh
+   kubectl label namespace default istio-injection=enabled
+   ```
+
+   1. Deploy the Bookinfo sample application to the default namespace, which you will expose behind a GraphQL server embedded in Envoy.
+   ```sh
+   kubectl apply -f samples/bookinfo/platform/kube/bookinfo.yaml
+   ```
+
+2. Verify that Gloo Edge automatically discovered the Bookinfo services and created corresponding `default-productpage-9080` upstream, which you will use in the REST resolver.
```
Comment (kcbabo) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:137]: Related to my comment above about FDS whitelist vs. blacklist, we probably don't want FDS and GraphQL autogeneration kicking in for any of the bookinfo upstreams. So more of a reason for us to constrain that earlier command.

@saiskee : would appreciate your thoughts here.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 12]
```diff
@@ -1,45 +1,75 @@
...
+
+## Step 2: GraphQL service discovery with Pet Store {#pet-store}
+
+Explore GraphQL service discovery with the Pet Store sample application.
+
+1. Start by deploying the Pet Store sample application, which you will expose behind a GraphQL server embedded in Envoy.
+   ```sh
+   kubectl apply -f https://raw.githubusercontent.com/solo-io/gloo/master/example/petstore/petstore.yaml
+   ```
+   Note that any `/GET` requests to `/api/pets` of this service return the following JSON output:
```
Comment (artberger) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:46]: Should there be a next step that has them make a GET request?

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 13]
```diff
@@ -1,45 +1,75 @@
...
+   kubectl get graphqlschemas -n gloo-system
+   ```
+
+   Example output:
+   ```
+   NAME                    AGE
+   default-petstore-8080   2m58s
+   ```
+
+4. Optional: Check out the generated GraphQL schema.
```
Comment (artberger) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:67]: Is there something in particular they should look for, or more that it just exists?

**Review by kcbabo — COMMENTED** [review]
This looks great, Rachael. I think you do a great job of capturing the two primary workflows with petstore and bookinfo. This is great for the initial writeup and we can add specifics around the resolver configuration as a follow-up.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 14]
```diff
@@ -52,57 +82,148 @@ spec:
...
+2. Verify that Gloo Edge automatically discovered the Bookinfo services and created corresponding `default-productpage-9080` upstream, which you will use in the REST resolver.
+```sh
+kubectl get upstream -n gloo-system
+```
+
+3. Check out the contents of the following Gloo Edge GraphQL schema CRD. Specifically, take a look at the `restResolver` and `schema_definition` sections.
+   ```sh
+   curl https://raw.githubusercontent.com/kcbabo/graphql-bookinfo/main/kubernetes/bookinfo-gql.yaml
+   ```
+   * `restResolver`: A resolver is defined by a name (ex: Query|productsForHome) and whether it is a REST or a gRPC resolver. This example is a REST resolver, so the path and the method that are needed to request the data are specified. The path can reference a parent attribute, such as `/details/{$parent.id}.`
```
Comment (artberger) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:146]:
```suggestion
   * `restResolver`: A resolver is defined by a name (ex: `Query|productsForHome`) and whether it is a REST or a gRPC resolver. This example is a REST resolver, so the path and the method that are needed to request the data are specified. The path can reference a parent attribute, such as `/details/{$parent.id}.`
```

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 15]
```diff
@@ -52,57 +82,148 @@ spec:
...
+         restResolver:
+           request:
+             headers:
+               :method: GET
+               :path: /api/v1/products
+           upstreamRef:
+             name: default-productpage-9080
+             namespace: gloo-system
+     ```
+   * `schema_definition`: A schema definition defines queries and types that reference the resolvers.
```
Comment (artberger) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:159]: I'm wondering how to use this...Does this mean that these are the way that users can filter the info returned from graphQL when they make a request to the endpoint? 

Or does this mean that here is where you have to set up how it uses the backing services? Or maybe both (like how you set up the backing services in this schema impacts what data the users can request).

If so, maybe rephrase it to point that out more.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 16]
```diff
@@ -52,57 +82,148 @@ spec:
...
 {{< /highlight >}}
 
-3. Send a request to the endpoint to verify that the request is successfully resolved by Envoy.
-   ```shell
-   curl "$(glooctl proxy url)/graphql" -H 'Content-Type: application/json' -d '{"query":"{pets{name}}"}'
+6. Send a request to the endpoint to verify that the request is successfully resolved by Envoy.
+   ```sh
+   curl "$(glooctl proxy url)/graphql" -H 'Content-Type: application/json' -d '{"query": "query {productsForHome {id, title, author, pages, year}}"}'
    ```
    Example successful response:
```
Comment (artberger) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:218]: Maybe something like, "Note that only the information you queried is returned like in the previous section?

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 13 above)
Comment (Rachael-Graham) [reply docs/content/guides/traffic_management/request_processing/graphql/_index.md:67]: nah they're just checking it out if they want to see the contents, but they dont need to do anything

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 1 above)
Comment (asayah) [reply docs/content/guides/traffic_management/request_processing/graphql/_index.md]: +1   1.11 has a diff api

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 7 above)
Comment (saiskee) [reply docs/content/guides/traffic_management/request_processing/graphql/_index.md:53]: I would append a note after saying something like:
Note: this turns on discovery for all upstreams. To use discovery with only certain upstreams, follow the FDS guide [here](https://docs.solo.io/gloo-edge/latest/installation/advanced_configuration/fds_mode/#function-discovery-service-fds).

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 11 above)
Comment (saiskee) [reply docs/content/guides/traffic_management/request_processing/graphql/_index.md:119]: This makes sense. I don't see an issue with fds kicking in for the bookinfo upstreams other than a bit of extra noise.  I don't think this is a blocker since the guide will still work and we can link them to the fds guide above to finely tune fds for certain upstreams.

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 17]
```diff
@@ -52,57 +82,135 @@ spec:
...
+   ```
+
+## Try it yourself
+
+You can deploy your own GraphQL API, which might not leverage automatic service discovery and registration.
+
+To get started, check out the in-depth example in the [`graphql-bookinfo` repository](https://github.com/kcbabo/graphql-bookinfo). You can model your own use case based on the contents of this example:
+* The `kubernetes` directory contains the Bookinfo sample app deployment, the example GraphQL schema, and the virtual service to route requests to the `/graphql` endpoint.
+* The `openapi` directory contains the OpenAPI specifications for the individual BookInfo microservices, along with the original consolidated BookInfo REST API.
+
```
Comment (Rachael-Graham) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:212]: @kcbabo added a link to your repo + some info about what is included so they can base their own app  and graphql schema off of this... not sure though if this covers your request fully for an explanation fo "workflow for deploying your own GraphQL API (i.e. one that doesn't leverage autodiscovery and registration)."

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 7 above)
Comment (Rachael-Graham) [reply docs/content/guides/traffic_management/request_processing/graphql/_index.md:53]: added below!

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md (hunk 11 above)
Comment (Rachael-Graham) [reply docs/content/guides/traffic_management/request_processing/graphql/_index.md:120]: done, see [line 55](https://github.com/solo-io/gloo/pull/5800/files#diff-ec88209c5e22e590489013249847e933989e7e2372efe815fa5a5ce9141e8c46R55)!

File: docs/content/guides/traffic_management/request_processing/graphql/_index.md [hunk 18]
```diff
@@ -1,45 +1,76 @@
...
+   Optional: You can [create a route and send a `/GET` request to `/api/pets` of this service]({{% versioned_link_path fromRoot="/guides/security/auth/custom_auth/#setup" %}}), which returns the following unfiltered JSON output:
+   ```json
+   [{"id":1,"name":"Dog","status":"available"},{"id":2,"name":"Cat","status":"pending"}]
+   ```
+
+2. To allow Gloo Edge to automatically discover API specifications and create GraphQL schemas, turn on FDS discovery.
+   ```sh
+   kubectl patch settings -n gloo-system default --type=merge --patch '{"spec":{"discovery":{"fdsMode":"BLACKLIST"}}}'
+   ```
+   Note that this setting enables discovery for all upstreams. To enable discovery for only specified upstreams, see the [Function Discovery Service (FDS) guide]({{% versioned_link_path fromRoot="/installation/advanced_configuration/fds_mode/#function-discovery-service-fds" %}}).
```
Comment (Rachael-Graham) [inline docs/content/guides/traffic_management/request_processing/graphql/_index.md:55]: Adding some notes to self from Adam's updated instruqt track

To allow Gloo Edge to automatically discover the APIs and create a GraphQL schema, we will use the WHITELIST FDS discovery mode and turn on discovery only for the Petstore service:

kubectl label service petstore discovery.solo.io/function_discovery=enabled

**Review by kcbabo — DISMISSED** [review]

**Review by artberger — APPROVED** [review]
